SonarQube.TeamBuild.PreProcessor/SonarLintAnalyzerProvider.cs

```diff
@@ -0,0 +1,139 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="SonarLintAnalyzerProvider.cs" company="SonarSource SA and Microsoft Corporation">
+//   Copyright (c) SonarSource SA and Microsoft Corporation.  All rights reserved.
+//   Licensed under the MIT License. See License.txt in the project root for license information.
+// </copyright>
+//-----------------------------------------------------------------------
+
+using SonarQube.Common;
+using SonarQube.TeamBuild.Integration;
+using System;
+using System.IO;
+using System.Linq;
+
+namespace SonarQube.TeamBuild.PreProcessor
+{
+    /// <summary>
+    /// Handles fetch the SonarLint ruleset and provisioning the SonarLint assemblies on the client
+    /// </summary>
+    /// <remarks>This code is a short-term solution introduced in MSBuild Scanner version 1.1.
+    /// Going forwards, we want to provide a more general solution that will work for all Roslyn
+    /// analyzers that the SonarQube server knows about. It is likely that the general-purpose
+    /// solution will be quite different.</remarks>
+    public static class SonarLintAnalyzerProvider
+    {
+        public const string RoslynCSharpRulesetFileName = "SonarQubeRoslyn-cs.ruleset";
+
+        public const string SonarLintProfileFormatName = "sonarlint-vs-cs";
+        public const string CSharpLanguage = "cs";
+        public const string CSharpPluginKey = "csharp";
+        public const string CSharpRepositoryKey = "csharp";
+
+        /// <summary>
+        /// Sets up the client to run the SonarLint analyzer as part of the build
+        /// i.e. creates the Rolsyn ruleset and provisions the analyzer assemblies
+        /// </summary>
+        public static void SetupAnalyzers(ISonarQubeServer server, TeamBuildSettings settings, string projectKey, ILogger logger)
+        {
+            if (server == null)
+            {
+                throw new ArgumentNullException("server");
+            }
+            if (settings == null)
+            {
+                throw new ArgumentNullException("settings");
+            }
+            if (string.IsNullOrWhiteSpace(projectKey))
+            {
+                throw new ArgumentNullException("projectKey");
+            }
+            if (logger == null)
+            {
+                throw new ArgumentNullException("logger");
+            }
+
+            if (IsCSharpPluginInstalled(server) &&
+                TryCreateRuleset(server, settings, projectKey, logger))
+            {
+                FetchBinaries(server, settings, logger);
+            }
+            else
+            {
+                logger.LogDebug(Resources.SLAP_CSharpPluginNotInstalled);
```

Comment:
If TryCreateRuleset fails (and there seem to be many error paths), the "C# plugin is not installed" message will be logged, which is not factual.